File: Template/Dijkstra.cpp
//? 다익스트라
#include <iostream>
#include <queue>
#include <vector>
#include <utility>
#include <algorithm>

typedef std::pair<int, int> pii;
const int MAX = 1001;
const int INF = 999999;
std::vector<pii> adj[MAX];
int dist[MAX], prev[MAX];

int func(int start, int end) {
    std::priority_queue<pii, std::vector<pii>, std::greater<pii> > pq;    
    std::fill(dist,dist+MAX,INF);
    std::fill(prev,prev+MAX,-1);
    dist[start] = 0;
    pq.push({dist[start], start});

    while(!pq.empty())
    {
        const auto &p = pq.top();
        int cost = p.first;
        int curr = p.second;
        pq.pop();
        if(dist[curr] < cost) continue;
        for(auto q : adj[curr]) {
            int next = q.first;
            int d = q.second + cost;
            if(d < dist[next])
            {
                dist[next] = d;
                prev[next] = curr;
                pq.push({dist[next], next});
            }
        }
    }

    return dist[end];
}

void track(int node, int step)
{
    if(prev[node] == -1) {
        std::cout << step << '\n';
        std::cout << node << ' ';
        return;
    }
    track(prev[node], step+1);
    std::cout << node << ' ';
}

int main()
{
    return 0;
}

File: Template/Python/Dijkstra.py
# https://www.acmicpc.net/problem/1753
import sys, heapq

prev = [-1 for _ in range(20_001)]
def track(node, step):
  if prev[node] == -1:
    print(step)
    print(node, end=' ')
    return
  track(prev[node], step+1)
  print(node, end=' ')

MAX = 1_000_000_000
V, E = map(int, sys.stdin.readline().split())
start = int(sys.stdin.readline())
adj = [[] for _ in range(20_001)]
dist = [MAX for _ in range(20_001)]
pq = []


for _ in range(E):
  u, v, w = map(int, sys.stdin.readline().split())
  adj[u].append((v, w))

heapq.heappush(pq, (0, start))

while pq:
  cost, curr = pq[0]
  heapq.heappop(pq)

  if dist[curr] < cost:
    continue

  for next, d in adj[curr]:
    d += cost
    if dist[next] > d:
      dist[next] = d
      prev[next] = curr
      heapq.heappush(pq, (dist[next], next))


for i in range(1, V+1):
  if i == start:
    print(0)
  else:
    print(("INF" if dist[i] == MAX else dist[i]))